bluetooth/src/commonMain/kotlin/device/DeviceState.kt

```diff
@@ -323,17 +290,209 @@ sealed class DeviceState(
+                            is ConnectionSettings.ReconnectionSettings.Never -> {
+                                clean()
+                                state.didDisconnect
+                            }
+                        }
+                        is DeviceState.Disconnected -> state.remain()
+                        is DeviceState.Connecting,
+                        is DeviceState.Disconnecting -> {
+                            clean()
+                            state.didDisconnect
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun monitorDiscovering() {
+        launch {
+            connectionManager.startedDiscovering.collect {
+                takeAndChangeState { deviceState ->
+                    if (deviceState is DeviceState.Connected.NoServices)
+                        deviceState.discoverServices
+                    else
+                        deviceState.remain()
+                }
+            }
+        }
+    }
+
+    private fun monitorDiscovered() {
+        launch {
+            connectionManager.discoverCompleted.collect { services ->
+                takeAndChangeState { state ->
+                    when (state) {
+                        is DeviceState.Connected.Discovering -> state.didDiscoverServices(services)
+                        else -> state.remain()
+                    }
+                }
+            }
+        }
+    }
+
+    private fun monitorNewActions() {
+        launch {
+            connectionManager.newAction.collect { action ->
+                takeAndChangeState { state ->
+                    when (state) {
+                        is DeviceState.Connected.Idle -> {
+                            state.handleAction(action)
+                        }
+                        is DeviceState.Connected.HandlingAction -> {
+                            state.addAction(action)
+                        }
+                        is DeviceState.Connected.NoServices,
+                        is DeviceState.Connected.Discovering,
+                        is DeviceState.Connecting,
+                        is DeviceState.Reconnecting,
+                        is DeviceState.Disconnected,
+                        is DeviceState.Disconnecting,
+                        -> {
+                            state.remain() // TODO consider an optional buffer
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun monitorActionCompleted() {
+        launch {
+            connectionManager.actionCompleted.collect { (action, succeeded) ->
+                takeAndChangeState { state ->
+                    if (state is DeviceState.Connected.HandlingAction && state.action == action) {
+                        state.action.completedSuccessfully.complete(succeeded)
+                        debug(TAG) { "Action $action has been succeeded: $succeeded" }
```

Comment:
rogue debug I think

Reply:
That one added by me and it's very handy during Bluetooth development

Reply:
We should probably add a logging flag to be passed to bluetooth so this can be disabled on production though. Will add that as part of the BT upgrade

Reply:
IMHO these should then be disabled by default, also see:

https://github.com/splendo/kaluga/issues/302

logging anything related to data (even just the metadata) is very sensitive and should be disabled by default.

Reply:
+ 1. Very useful. And I'm okay to keep it here as long as it has a TAG

Reply:
It should not be too hard to implement it along the lines of #302 but it needs to be off by default

Reply:
Yep, doing it as part of BT rewrite